File: mns-reactor/src/main/java/cz/fb/manaus/reactor/charge/MarketCharge.java
package cz.fb.manaus.reactor.charge;

import com.google.common.collect.ImmutableMap;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;

import java.util.HashMap;
import java.util.Map;

import static java.lang.Math.max;

public class MarketCharge {

    private final double totalProfit;
    private final double totalPositiveProfit;
    private final double totalCharge;
    private final Map<String, Double> profits;

    private MarketCharge(double totalProfit, double totalPositiveProfit, double totalCharge, Map<String, Double> profits) {
        this.totalProfit = totalProfit;
        this.totalPositiveProfit = totalPositiveProfit;
        this.totalCharge = totalCharge;
        this.profits = ImmutableMap.copyOf(profits);
    }

    public static MarketCharge fromBets(double chargeRate, Iterable<SettledBet> bets) {
        var profits = new HashMap<String, Double>();
        double totalProfit = 0d, totalPositiveProfit = 0d;
        for (var bet : bets) {
            var betId = bet.getBetAction().getBetId();
            profits.put(betId, bet.getProfitAndLoss());
            totalProfit += bet.getProfitAndLoss();
            totalPositiveProfit += max(bet.getProfitAndLoss(), 0d);
        }
        var totalCharge = Price.round(chargeRate * max(totalProfit, 0));
        return new MarketCharge(totalProfit, totalPositiveProfit, totalCharge, profits);
    }

    public double getChargeContribution(String betId) {
        if (Price.amountEq(totalCharge, 0d)) return 0d;
        var profit = max(profits.get(betId), 0d);
        return Price.round(totalCharge * profit / totalPositiveProfit);
    }

    public double getTotalProfit() {
        return Price.round(totalProfit);
    }

    public double getTotalCharge() {
        return Price.round(totalCharge);
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/charge/MarketCharge.kt
package cz.fb.manaus.reactor.charge

import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.SettledBet
import cz.fb.manaus.core.model.amountEq
import kotlin.math.max

data class MarketCharge(private val totalProfit: Double,
                        private val totalPositiveProfit: Double,
                        private val totalCharge: Double,
                        private val profits: Map<String, Double>) {

    fun getChargeContribution(betId: String): Double {
        if (totalCharge amountEq 0.0) return 0.0
        val profit = max(profits[betId]!!, 0.0)
        return Price.round(totalCharge * profit / totalPositiveProfit)
    }

    fun getTotalProfit(): Double {
        return Price.round(totalProfit)
    }

    fun getTotalCharge(): Double {
        return Price.round(totalCharge)
    }

    companion object {
        fun fromBets(chargeRate: Double, bets: Iterable<SettledBet>): MarketCharge {
            val profits = bets.associate { it.id to it.profitAndLoss }.toMap()
            val totalProfit = bets.sumByDouble { it.profitAndLoss }
            val totalPositiveProfit = bets.map { it.profitAndLoss }.filter { it > 0 }.sum()
            val totalCharge = Price.round(chargeRate * max(totalProfit, 0.0))
            return MarketCharge(totalProfit, totalPositiveProfit, totalCharge, profits)
        }
    }
}
